pep-0727.rst

```diff
@@ -0,0 +1,516 @@
+    ) -> Annotated[User, doc("The created user after saving in the database")]:
+        """
+        Create a new user in the system, it needs the database connection to be already
+        initialized.
+        """
+        pass
+
+
+The return of the ``doc()`` function is an instance of a class that can be checked
+and used at runtime, defined similar to:
+
+.. code-block::
+
+    class DocInfo:
+        def __init__(self, documentation: str):
+            self.documentation = documentation
+
+...where the attribute ``documentation`` contains the same value string passed to
+the function ``doc()``.
+
+
+Additional Scenarios
+--------------------
+
+The main scenarios that this proposal intends to cover are described above, and
+for implementers to be conformant to this specification, they only need to support
+those scenarios described above.
+
+Here are some additional edge case scenarios with their respective considerations,
+but implementers are not required to support them.
+
+
+Type Alias
+----------
+
+When creating a type alias, like:
+
+.. code-block::
+
+    Username = Annotated[str, doc("The name of a user in the system")]
+
+
+...the documentation would be considered to be carried by the parameter annotated
+with ``Username``.
+
+So, in a function like:
+
+.. code-block::
+
+    def hi(
+      to: Username,
+    ) -> None: ...
+
+
+...it would be equivalent to:
+
+.. code-block::
+
+    def hi(
+      to: Annotated[str, doc("The name of a user in the system")],
+    ) -> None: ...
+
+Nevertheless, implementers would not be required to support type aliases outside
+of the final type annotation to be conformant with this specification, as it
+could require more complex dereferencing logic.
+
+
+Annotating Type Parameters
+--------------------------
+
+When annotating type parameters, as in:
+
+.. code-block::
+
+    def hi(
+      to: list[Annotated[str, doc("The name of a user in a list")]],
+    ) -> None: ...
+
+...the documentation in ``doc()`` would refer to what it is annotating, in this
+case, each item in the list, not the list itself.
+
+There are currently no practical use cases for documenting type parameters,
+so implementers are not required to support this scenario to be considered
+conformant, but it's included for completeness.
+
+
+Annotating Unions
+-----------------
+
+If used in one of the parameters of a union, as in:
+
+.. code-block::
+
+    def hi(
+      to: str | Annotated[list[str], doc("List of user names")],
+    ) -> None: ...
+
+...again, the documentation in ``doc()`` would refer to what it is annotating,
+in this case, this documents the list itself, not its items.
+
+In particular, the documentation would not refer to a single string passed as a
+parameter, only to a list.
+
+There are currently no practical use cases for documenting unions, so implementers
+are not required to support this scenario to be considered conformant, but it's
+included for completeness.
```

Comment:
Again opinion, but I think we should be more conservative in the initial implementation and ban these things, if we think they don't make sense (I'm trying to work out what on earth I'd do here in Sphinx)